src/python/WMCore/WorkQueue/WorkQueueBackend.py

```diff
@@ -321,112 +321,127 @@ def deleteElements(self, *elements):
-
-        sites = thresholds.keys()
-        self.logger.info("Current siteJobCounts:")
-        for site, jobsByPrio in siteJobCounts.items():
-            self.logger.info("    %s : %s", site, jobsByPrio)
-
-        for element in sortedElements:
-            commonSites = possibleSites(element)
-            prio = element['Priority']
-            possibleSite = None
-            random.shuffle(sites)
-            for site in sites:
-                if site in commonSites:
-                    # Count the number of jobs currently running of greater priority
-                    curJobCount = sum([x[1] if x[0] >= prio else 0 for x in siteJobCounts.get(site, {}).items()])
-                    self.logger.debug("Job Count: %s, site: %s thresholds: %s" % (curJobCount, site, thresholds[site]))
-                    if curJobCount < thresholds[site]:
-                        possibleSite = site
-                        break
-
-            if possibleSite:
-                elements.append(element)
-                siteJobCounts.setdefault(possibleSite, {})
-                siteJobCounts[possibleSite][prio] = siteJobCounts[possibleSite].setdefault(prio, 0) + \
-                                                    element['Jobs'] * element.get('blowupFactor', 1.0)
             else:
-                self.logger.debug("No available resources for %s with doc id %s", element['RequestName'], element.id)
+                self.logger.info("All the workqueue elements have been exhausted for: %s ", self.queueUrl)
+                break
+            # update number of documents to skip in the next cycle
+            numSkip += docsSliceSize
+
+            # Convert python dictionary into Couch WQE objects, skipping aborted workflows
+            # And sort them by creation time and priority, such that highest priority and
+            # oldest elements come first in the list
+            sortedElements = []
+            for i in result:
+                element = CouchWorkQueueElement.fromDocument(self.db, i)
+                # make sure not to acquire work for aborted or force-completed workflows
+                if element['RequestName'] in excludeWorkflows:
+                    msg = "Skipping aborted/force-completed workflow: %s, work id: %s"
+                    self.logger.info(msg, element['RequestName'], element._id)
+                else:
+                    sortedElements.append(element)
+            sortedElements.sort(key=lambda element: element['CreationTime'])
+            sortedElements.sort(key=lambda element: element['Priority'], reverse=True)
+
+            for element in sortedElements:
+                if numElems <= 0:
+                    msg = "Reached maximum number of elements to be accepted, "
+                    msg += "configured to: {}, from queue: {}".format(len(elements), self.queueUrl)
+                    self.logger.info(msg)
+                    breakOut = True  # get out of the outer loop as well
```

Comment:
One question here:  Do you account for or is there some other kind of protection for the fact that if you have exhausted the "Maximum number of elements  to accept" counter while you are in the middle of the current slice of elements fetched from the Queue, then on your next run you are about to skip the rest of the elements that are in the current slice and not yet considered at all. I am not sure this would ever be a problem, though.

Reply:
Yes, we break out of this loop right away and the first thing to be checked in the `while` loop is whether this flag `breakOut` is True or not. If true, then the whole loop going through the elements is over.